connectors/sources/redis.py

```diff
@@ -0,0 +1,283 @@
+                elif invalid_type:
+                    msg = f"All database element should be integer. Please correct database name: {', '.join(invalid_type)}"
+                if msg:
+                    raise ConfigurableFieldValueError(msg)
+
+    async def validate_config(self):
+        """Validates whether user input is empty or not for configuration fields
+        Also validate, if user configured databases are available in Redis."""
+        await super().validate_config()
+        await self._remote_validation()
+
+    async def format_document(self, **kwargs):
+        """Prepare document for database records.
+
+        Returns:
+            document: Modified document.
+        """
+        doc_id = hash_id(f"{kwargs.get('db')}/{kwargs.get('key')}")
+        document = {
+            "_id": doc_id,
+            "key": kwargs.get("key"),
+            "value": str(kwargs.get("value")),
+            "size_in_bytes": kwargs.get("size"),
+            "database": kwargs.get("db"),
+            "key_type": kwargs.get("key_type"),
+            "_timestamp": iso_utc(),
+        }
+        return document
+
+    async def ping(self):
+        try:
+            async with self.redis_client.client() as redis_client:
+                await redis_client.ping()
+                self._logger.info("Successfully connected to Redis.")
+        except Exception:
+            self._logger.exception("Error while connecting to Redis.")
+            raise
+
+    async def get_key_metadata(self, redis_client, key, db):
+        key_type = await redis_client.type(key)
+        key_value = await self.redis_client.get_key_value(
+            redis_client=redis_client, key=key, key_type=key_type
+        )
+        key_size = await redis_client.memory_usage(key)
+        return await self.format_document(
+            key=key,
+            value=key_value,
+            key_type=key_type,
+            size=key_size,
+            db=db,
+        )
+
+    async def get_db_records(self, db, redis_client, pattern="*", _type=None):
+        await redis_client.execute_command("SELECT", db)
+        async for keys in self.redis_client._get_paginated_keys(
+            redis_client=redis_client, pattern=pattern, _type=_type
+        ):
+            for key in keys:
+                document = await self.get_key_metadata(
+                    redis_client=redis_client, key=key, db=db
+                )
+                yield document
+
+    async def get_docs(self, filtering=None):
+        """Get documents from Redis
+
+        Returns:
+            dictionary: Document of database content
+
+        Yields:
+            dictionary: Document from Redis.
+        """
+        async with self.redis_client.client() as redis_client:
+            db_list = await self.redis_client.get_databases()
+            for db in db_list:
+                async for document in self.get_db_records(
+                    db=db, redis_client=redis_client
```

Comment:
It feels strange that you need to pass `redis_client` to almost every method of `self.redis_client`, while `redis_client` is from `self.redis_client`.

I think this kind of details should be hidden inside `self.redis_client`.

Reply:
`redis_client` is a context manager if we store it in `self.redis_client` then it will get close everytime.

Reply:
Do you want to keep the pool open or the redis client as well?

Reply:
Yes, I guess we need to open the pool/client until all documents get index into elastic search. Otherwise we need to create it every time which degrade the performance of connector. We also faced the performance issue in S3 because of creating multiple clients.

Reply:
I also see this is used inside `RedisClient`:

https://github.com/elastic/connectors/blob/4147a65dac50e0b6611d52b3f1e7f1e703b1994f/connectors/sources/redis.py#L75

if you only want to have one instance of the client and keep it open until the end of the `RedisClient`, consider assigning it to an instance variable, and close it in a `close` method.

Reply:
As of now we didn't added code for ssl support. But with ssl we need to pass the path of file in query while creating client as we don't have a way to pass ssl context and In that case we'll face the same issue as mongodb.
Please refer this issue for better understanding: https://github.com/elastic/connectors/issues/1948

Reply:
From the code snippet you posted, It seems the ssl path was passed when creating the pool. So ideally, we should have a lazy instantiated pool in `RedisClient` instance, and we should close it with a `close` method. From the pool, you can create a redis client via `redis.Redis.from_pool(pool)`.

>and in the call of the `validate`, connector was never called `close`.

I think it's the responsibility of the users (of `RedisClient`) to `close` it after use.

Reply:
Just curious, must a Redis client instance be instantiated from a pool? I don't think we have the use case of multiple connections.

Reply:
I think from the comment, Artem mentioned 
>call of `validate` on connector never called `close` after validation is ended.

But it doesn't mean the temporary files cannot be removed from `close` method.